File: src/go/session.go
package geminisdk

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"

	"github.com/google/uuid"
)

// ToolHandler is a function that handles tool invocations
type ToolHandler func(ctx context.Context, invocation ToolInvocation) ToolResult

// EventHandler is a function that handles session events
type EventHandler func(event SessionEvent)

// Session represents a conversation session
type Session struct {
	sessionID        string
	model            string
	backend          *Backend
	tools            []Tool
	toolHandlers     map[string]ToolHandler
	systemMessage    string
	generationConfig *GenerationConfig
	thinkingConfig   *ThinkingConfig
	streaming        bool

	messages      []Message
	eventHandlers []EventHandler
	closed        bool
	startTime     time.Time
	modifiedTime  time.Time
	mu            sync.RWMutex
}

// NewSession creates a new session
func NewSession(
	sessionID string,
	model string,
	backend *Backend,
	tools []Tool,
	systemMessage string,
	generationConfig *GenerationConfig,
	thinkingConfig *ThinkingConfig,
	streaming bool,
) *Session {
	now := time.Now()
	s := &Session{
		sessionID:        sessionID,
		model:            model,
		backend:          backend,
		tools:            tools,
		toolHandlers:     make(map[string]ToolHandler),
		systemMessage:    systemMessage,
		generationConfig: generationConfig,
		thinkingConfig:   thinkingConfig,
		streaming:        streaming,
		messages:         make([]Message, 0),
		eventHandlers:    make([]EventHandler, 0),
		closed:           false,
		startTime:        now,
		modifiedTime:     now,
	}

	if systemMessage != "" {
		s.messages = append(s.messages, Message{
			Role:    RoleSystem,
			Content: systemMessage,
		})
	}

	return s
}

// SessionID returns the session ID
func (s *Session) SessionID() string {
	return s.sessionID
}

// Model returns the model name
func (s *Session) Model() string {
	return s.model
}

// StartTime returns when the session started
func (s *Session) StartTime() time.Time {
	return s.startTime
}

// ModifiedTime returns when the session was last modified
func (s *Session) ModifiedTime() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.modifiedTime
}

// Messages returns the conversation history
func (s *Session) Messages() []Message {
	s.mu.RLock()
	defer s.mu.RUnlock()
	result := make([]Message, len(s.messages))
	copy(result, s.messages)
	return result
}

// RegisterToolHandler registers a handler for a tool
func (s *Session) RegisterToolHandler(name string, handler ToolHandler) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.toolHandlers[name] = handler
}

// On registers an event handler
func (s *Session) On(handler EventHandler) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.eventHandlers = append(s.eventHandlers, handler)
}

func (s *Session) emit(eventType EventType, data map[string]interface{}) {
	event := SessionEvent{
		EventType: eventType,
		Data:      data,
		SessionID: s.sessionID,
	}

	s.mu.RLock()
	handlers := make([]EventHandler, len(s.eventHandlers))
	copy(handlers, s.eventHandlers)
	s.mu.RUnlock()

	for _, handler := range handlers {
		handler(event)
	}
}

// Send sends a message to the session
func (s *Session) Send(ctx context.Context, options *MessageOptions) error {
	s.mu.RLock()
	if s.closed {
		s.mu.RUnlock()
		return NewSessionClosedError(s.sessionID)
	}
	s.mu.RUnlock()

	content := options.Prompt
	if options.Context != "" {
		content = options.Context + "\n\n" + content
	}

	userMessage := Message{
		Role:    RoleUser,
		Content: content,
	}

	s.mu.Lock()
	s.messages = append(s.messages, userMessage)
	s.modifiedTime = time.Now()
	s.mu.Unlock()

	var err error
	if s.streaming {
		err = s.streamResponse(ctx)
	} else {
		err = s.getResponse(ctx)
	}

	if err != nil {
		s.emit(EventSessionError, map[string]interface{}{"error": err.Error()})
	}

	return err
}

// SendAndWait sends a message and waits for the response
func (s *Session) SendAndWait(ctx context.Context, options *MessageOptions) (*SessionEvent, error) {
	resultCh := make(chan *SessionEvent, 1)
	errCh := make(chan error, 1)

	handler := func(event SessionEvent) {
		switch event.EventType {
		case EventAssistantMessage, EventSessionIdle, EventSessionError:
			select {
			case resultCh <- &event:
			default:
			}
		}
	}

	s.On(handler)

	if err := s.Send(ctx, options); err != nil {
		return nil, err
	}

	select {
	case event := <-resultCh:
		return event, nil
	case err := <-errCh:
		return nil, err
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

func (s *Session) streamResponse(ctx context.Context) error {
	var fullContent string
	var fullReasoning string
	var allToolCalls []ToolCall
	var finalUsage *LLMUsage

	s.mu.RLock()
	messages := make([]Message, len(s.messages))
	copy(messages, s.messages)
	tools := s.tools
	s.mu.RUnlock()

	stream, err := s.backend.CompleteStreaming(
		ctx,
		s.model,
		messages,
		s.generationConfig,
		s.thinkingConfig,
		tools,
	)
	if err != nil {
		return err
	}

	for result := range stream {
		if result.Error != nil {
			return result.Error
		}

		chunk := result.Chunk
		if chunk.Content != "" {
			fullContent += chunk.Content
			s.emit(EventAssistantMessageDelta, map[string]interface{}{
				"deltaContent": chunk.Content,
				"content":      fullContent,
			})
		}

		if chunk.ReasoningContent != "" {
			fullReasoning += chunk.ReasoningContent
			s.emit(EventAssistantReasoningDelta, map[string]interface{}{
				"deltaContent": chunk.ReasoningContent,
				"content":      fullReasoning,
			})
		}

		if len(chunk.ToolCalls) > 0 {
			allToolCalls = append(allToolCalls, chunk.ToolCalls...)
		}

		if chunk.Usage != nil {
			finalUsage = chunk.Usage
		}
	}

	if len(allToolCalls) > 0 {
		if err := s.handleToolCalls(ctx, allToolCalls); err != nil {
			return err
		}
	}

	assistantMessage := Message{
		Role:      RoleAssistant,
		Content:   fullContent,
		ToolCalls: allToolCalls,
	}

	s.mu.Lock()
	s.messages = append(s.messages, assistantMessage)
	s.mu.Unlock()

	if fullReasoning != "" {
		s.emit(EventAssistantReasoning, map[string]interface{}{
			"content": fullReasoning,
		})
	}

	s.emit(EventAssistantMessage, map[string]interface{}{
		"content":   fullContent,
		"toolCalls": allToolCalls,
		"usage":     finalUsage,
	})

	s.emit(EventSessionIdle, map[string]interface{}{})

	return nil
}

func (s *Session) getResponse(ctx context.Context) error {
	s.mu.RLock()
	messages := make([]Message, len(s.messages))
	copy(messages, s.messages)
	tools := s.tools
	s.mu.RUnlock()

	chunk, err := s.backend.Complete(
		ctx,
		s.model,
		messages,
		s.generationConfig,
		s.thinkingConfig,
		tools,
	)
	if err != nil {
		return err
	}

	if len(chunk.ToolCalls) > 0 {
		if err := s.handleToolCalls(ctx, chunk.ToolCalls); err != nil {
			return err
		}
	}

	assistantMessage := Message{
		Role:      RoleAssistant,
		Content:   chunk.Content,
		ToolCalls: chunk.ToolCalls,
	}

	s.mu.Lock()
	s.messages = append(s.messages, assistantMessage)
	s.mu.Unlock()

	if chunk.ReasoningContent != "" {
		s.emit(EventAssistantReasoning, map[string]interface{}{
			"content": chunk.ReasoningContent,
		})
	}

	s.emit(EventAssistantMessage, map[string]interface{}{
		"content":   chunk.Content,
		"toolCalls": chunk.ToolCalls,
		"usage":     chunk.Usage,
	})

	s.emit(EventSessionIdle, map[string]interface{}{})

	return nil
}

func (s *Session) handleToolCalls(ctx context.Context, toolCalls []ToolCall) error {
	for _, tc := range toolCalls {
		toolName := tc.Function.Name

		s.emit(EventToolCall, map[string]interface{}{
			"name":      toolName,
			"arguments": json.RawMessage(tc.Function.Arguments),
			"callId":    tc.ID,
		})

		s.mu.RLock()
		handler, ok := s.toolHandlers[toolName]
		s.mu.RUnlock()

		if !ok {
			fmt.Printf("Warning: No handler for tool: %s\n", toolName)
			s.mu.Lock()
			s.messages = append(s.messages, Message{
				Role:       RoleUser,
				Content:    fmt.Sprintf("Error: Tool '%s' not found", toolName),
				Name:       toolName,
				ToolCallID: tc.ID,
			})
			s.mu.Unlock()
			continue
		}

		var args map[string]interface{}
		if err := json.Unmarshal(tc.Function.Arguments, &args); err != nil {
			args = make(map[string]interface{})
		}

		invocation := ToolInvocation{
			Name:      toolName,
			Arguments: args,
			CallID:    tc.ID,
		}

		result := handler(ctx, invocation)
		resultText := result.TextResultForLLM
		if resultText == "" {
			resultText = "Success"
		}

		s.emit(EventToolResult, map[string]interface{}{
			"name":   toolName,
			"callId": tc.ID,
			"result": resultText,
		})

		s.mu.Lock()
		s.messages = append(s.messages, Message{
			Role:       RoleUser,
			Content:    resultText,
			Name:       toolName,
			ToolCallID: tc.ID,
		})
		s.mu.Unlock()
	}

	return nil
}

// AddTool adds a tool to the session
func (s *Session) AddTool(tool Tool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.tools = append(s.tools, tool)
}

// RemoveTool removes a tool from the session
func (s *Session) RemoveTool(toolName string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	newTools := make([]Tool, 0, len(s.tools))
	for _, t := range s.tools {
		if t.Name != toolName {
			newTools = append(newTools, t)
		}
	}
	s.tools = newTools
	delete(s.toolHandlers, toolName)
}

// ClearHistory clears the conversation history
func (s *Session) ClearHistory() {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.messages = make([]Message, 0)
	if s.systemMessage != "" {
		s.messages = append(s.messages, Message{
			Role:    RoleSystem,
			Content: s.systemMessage,
		})
	}
	s.modifiedTime = time.Now()
}

// Destroy closes and cleans up the session
func (s *Session) Destroy() {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.closed = true
	s.eventHandlers = nil
	s.messages = nil
}

// CreateTool helper function to create a tool
func CreateTool(name, description string, parameters map[string]interface{}) Tool {
	var paramsJSON json.RawMessage
	if parameters != nil {
		paramsJSON, _ = json.Marshal(parameters)
	}
	return Tool{
		Name:        name,
		Description: description,
		Parameters:  paramsJSON,
	}
}

// SuccessResult creates a success tool result
func SuccessResult(text string) ToolResult {
	return ToolResult{
		ResultType:       ToolResultSuccess,
		TextResultForLLM: text,
	}
}

// FailureResult creates a failure tool result
func FailureResult(text string) ToolResult {
	return ToolResult{
		ResultType:       ToolResultFailure,
		TextResultForLLM: text,
	}
}

// RejectedResult creates a rejected tool result
func RejectedResult(text string) ToolResult {
	return ToolResult{
		ResultType:       ToolResultRejected,
		TextResultForLLM: text,
	}
}

// GenerateSessionID generates a new session ID
func GenerateSessionID() string {
	return uuid.New().String()
}

File: src/cpp/src/session.cpp
/**
 * @file session.cpp
 * @brief Session implementation for GeminiSDK C++
 */

#include "geminisdk/session.hpp"
#include "geminisdk/errors.hpp"
#include <iostream>

namespace geminisdk {

Session::Session(
    const std::string& session_id,
    const std::string& model,
    std::shared_ptr<Backend> backend,
    const std::vector<Tool>& tools,
    const std::optional<std::string>& system_message,
    const std::optional<GenerationConfig>& generation_config,
    const std::optional<ThinkingConfig>& thinking_config,
    bool streaming
) : session_id_(session_id),
    model_(model),
    backend_(backend),
    tools_(tools),
    system_message_(system_message),
    generation_config_(generation_config),
    thinking_config_(thinking_config),
    streaming_(streaming),
    closed_(false),
    start_time_(std::chrono::system_clock::now()),
    modified_time_(std::chrono::system_clock::now()) {
    
    if (system_message_.has_value()) {
        Message sys_msg;
        sys_msg.role = Role::System;
        sys_msg.content = *system_message_;
        messages_.push_back(sys_msg);
    }
}

Session::~Session() {
    destroy();
}

std::chrono::system_clock::time_point Session::modified_time() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return modified_time_;
}

std::vector<Message> Session::messages() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return messages_;
}

void Session::register_tool_handler(const std::string& name, ToolHandler handler) {
    std::lock_guard<std::mutex> lock(mutex_);
    tool_handlers_[name] = handler;
}

void Session::on(EventHandler handler) {
    std::lock_guard<std::mutex> lock(mutex_);
    event_handlers_.push_back(handler);
}

void Session::emit(EventType event_type, const json& data) {
    SessionEvent event;
    event.event_type = event_type;
    event.data = data;
    event.session_id = session_id_;
    
    std::vector<EventHandler> handlers;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        handlers = event_handlers_;
    }
    
    for (const auto& handler : handlers) {
        handler(event);
    }
}

void Session::send(const MessageOptions& options) {
    {
        std::lock_guard<std::mutex> lock(mutex_);
        if (closed_) {
            throw SessionClosedError(session_id_);
        }
    }
    
    std::string content = options.prompt;
    if (options.context.has_value()) {
        content = *options.context + "\n\n" + content;
    }
    
    Message user_message;
    user_message.role = Role::User;
    user_message.content = content;
    
    {
        std::lock_guard<std::mutex> lock(mutex_);
        messages_.push_back(user_message);
        modified_time_ = std::chrono::system_clock::now();
    }
    
    try {
        if (streaming_) {
            stream_response();
        } else {
            get_response();
        }
    } catch (const std::exception& e) {
        emit(EventType::SessionError, {{"error", e.what()}});
        throw;
    }
}

SessionEvent Session::send_and_wait(const MessageOptions& options) {
    SessionEvent result;
    bool received = false;
    std::mutex result_mutex;
    std::condition_variable cv;
    
    auto handler = [&](const SessionEvent& event) {
        if (event.event_type == EventType::AssistantMessage ||
            event.event_type == EventType::SessionIdle ||
            event.event_type == EventType::SessionError) {
            std::lock_guard<std::mutex> lock(result_mutex);
            if (!received) {
                result = event;
                received = true;
                cv.notify_one();
            }
        }
    };
    
    on(handler);
    send(options);
    
    std::unique_lock<std::mutex> lock(result_mutex);
    cv.wait(lock, [&] { return received; });
    
    return result;
}

void Session::stream_response() {
    std::string full_content;
    std::string full_reasoning;
    std::vector<ToolCall> all_tool_calls;
    std::optional<LLMUsage> final_usage;
    
    std::vector<Message> msgs;
    std::vector<Tool> current_tools;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        msgs = messages_;
        current_tools = tools_;
    }
    
    backend_->complete_streaming(
        model_,
        msgs,
        [&](const LLMChunk& chunk) {
            if (!chunk.content.empty()) {
                full_content += chunk.content;
                emit(EventType::AssistantMessageDelta, {
                    {"deltaContent", chunk.content},
                    {"content", full_content}
                });
            }
            
            if (chunk.reasoning_content.has_value()) {
                full_reasoning += *chunk.reasoning_content;
                emit(EventType::AssistantReasoningDelta, {
                    {"deltaContent", *chunk.reasoning_content},
                    {"content", full_reasoning}
                });
            }
            
            for (const auto& tc : chunk.tool_calls) {
                all_tool_calls.push_back(tc);
            }
            
            if (chunk.usage.has_value()) {
                final_usage = chunk.usage;
            }
        },
        generation_config_,
        thinking_config_,
        current_tools
    );
    
    if (!all_tool_calls.empty()) {
        handle_tool_calls(all_tool_calls);
    }
    
    Message assistant_message;
    assistant_message.role = Role::Assistant;
    assistant_message.content = full_content;
    assistant_message.tool_calls = all_tool_calls;
    
    {
        std::lock_guard<std::mutex> lock(mutex_);
        messages_.push_back(assistant_message);
    }
    
    if (!full_reasoning.empty()) {
        emit(EventType::AssistantReasoning, {{"content", full_reasoning}});
    }
    
    json msg_data = {{"content", full_content}};
    if (!all_tool_calls.empty()) {
        json tc_array = json::array();
        for (const auto& tc : all_tool_calls) {
            tc_array.push_back({
                {"id", tc.id},
                {"type", tc.type},
                {"function", {
                    {"name", tc.function.name},
                    {"arguments", tc.function.arguments}
                }}
            });
        }
        msg_data["toolCalls"] = tc_array;
    }
    if (final_usage.has_value()) {
        msg_data["usage"] = {
            {"promptTokens", final_usage->prompt_tokens},
            {"completionTokens", final_usage->completion_tokens},
            {"totalTokens", final_usage->total_tokens}
        };
    }
    
    emit(EventType::AssistantMessage, msg_data);
    emit(EventType::SessionIdle, {});
}

void Session::get_response() {
    std::vector<Message> msgs;
    std::vector<Tool> current_tools;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        msgs = messages_;
        current_tools = tools_;
    }
    
    LLMChunk chunk = backend_->complete(
        model_,
        msgs,
        generation_config_,
        thinking_config_,
        current_tools
    );
    
    if (!chunk.tool_calls.empty()) {
        handle_tool_calls(chunk.tool_calls);
    }
    
    Message assistant_message;
    assistant_message.role = Role::Assistant;
    assistant_message.content = chunk.content;
    assistant_message.tool_calls = chunk.tool_calls;
    
    {
        std::lock_guard<std::mutex> lock(mutex_);
        messages_.push_back(assistant_message);
    }
    
    if (chunk.reasoning_content.has_value()) {
        emit(EventType::AssistantReasoning, {{"content", *chunk.reasoning_content}});
    }
    
    json msg_data = {{"content", chunk.content}};
    if (!chunk.tool_calls.empty()) {
        json tc_array = json::array();
        for (const auto& tc : chunk.tool_calls) {
            tc_array.push_back({
                {"id", tc.id},
                {"type", tc.type},
                {"function", {
                    {"name", tc.function.name},
                    {"arguments", tc.function.arguments}
                }}
            });
        }
        msg_data["toolCalls"] = tc_array;
    }
    if (chunk.usage.has_value()) {
        msg_data["usage"] = {
            {"promptTokens", chunk.usage->prompt_tokens},
            {"completionTokens", chunk.usage->completion_tokens},
            {"totalTokens", chunk.usage->total_tokens}
        };
    }
    
    emit(EventType::AssistantMessage, msg_data);
    emit(EventType::SessionIdle, {});
}

void Session::handle_tool_calls(const std::vector<ToolCall>& tool_calls) {
    for (const auto& tc : tool_calls) {
        const std::string& tool_name = tc.function.name;
        
        emit(EventType::ToolCall, {
            {"name", tool_name},
            {"arguments", tc.function.arguments},
            {"callId", tc.id}
        });
        
        ToolHandler handler;
        {
            std::lock_guard<std::mutex> lock(mutex_);
            auto it = tool_handlers_.find(tool_name);
            if (it == tool_handlers_.end()) {
                std::cerr << "Warning: No handler for tool: " << tool_name << std::endl;
                Message error_msg;
                error_msg.role = Role::User;
                error_msg.content = "Error: Tool '" + tool_name + "' not found";
                error_msg.name = tool_name;
                error_msg.tool_call_id = tc.id;
                messages_.push_back(error_msg);
                continue;
            }
            handler = it->second;
        }
        
        ToolInvocation invocation;
        invocation.name = tool_name;
        invocation.call_id = tc.id;
        
        if (tc.function.arguments.is_object()) {
            for (auto it = tc.function.arguments.begin(); it != tc.function.arguments.end(); ++it) {
                invocation.arguments[it.key()] = it.value();
            }
        }
        
        try {
            ToolResult result = handler(invocation);
            std::string result_text = result.text_result_for_llm.value_or("Success");
            
            emit(EventType::ToolResult, {
                {"name", tool_name},
                {"callId", tc.id},
                {"result", result_text}
            });
            
            Message result_msg;
            result_msg.role = Role::User;
            result_msg.content = result_text;
            result_msg.name = tool_name;
            result_msg.tool_call_id = tc.id;
            
            std::lock_guard<std::mutex> lock(mutex_);
            messages_.push_back(result_msg);
        } catch (const std::exception& e) {
            std::string error_msg_str = "Error executing tool '" + tool_name + "': " + e.what();
            std::cerr << error_msg_str << std::endl;
            
            emit(EventType::ToolResult, {
                {"name", tool_name},
                {"callId", tc.id},
                {"error", error_msg_str}
            });
            
            Message result_msg;
            result_msg.role = Role::User;
            result_msg.content = error_msg_str;
            result_msg.name = tool_name;
            result_msg.tool_call_id = tc.id;
            
            std::lock_guard<std::mutex> lock(mutex_);
            messages_.push_back(result_msg);
        }
    }
}

void Session::add_tool(const Tool& tool) {
    std::lock_guard<std::mutex> lock(mutex_);
    tools_.push_back(tool);
}

void Session::remove_tool(const std::string& tool_name) {
    std::lock_guard<std::mutex> lock(mutex_);
    tools_.erase(
        std::remove_if(tools_.begin(), tools_.end(),
            [&](const Tool& t) { return t.name == tool_name; }),
        tools_.end()
    );
    tool_handlers_.erase(tool_name);
}

void Session::clear_history() {
    std::lock_guard<std::mutex> lock(mutex_);
    messages_.clear();
    
    if (system_message_.has_value()) {
        Message sys_msg;
        sys_msg.role = Role::System;
        sys_msg.content = *system_message_;
        messages_.push_back(sys_msg);
    }
    
    modified_time_ = std::chrono::system_clock::now();
}

void Session::destroy() {
    std::lock_guard<std::mutex> lock(mutex_);
    closed_ = true;
    event_handlers_.clear();
    messages_.clear();
}

} // namespace geminisdk
